### PR TITLE
Add auth actions and docs

### PR DESCRIPTION
**Pull request by kelkawi-a**
This PR adds instructions for enabling authentication and authorization. Authentication can be enabled through the Temporal web UI, Temporal worker and Temporal client libraries. Authorization (still a WIP based on this [PR](https://github.com/canonical/charmed-temporal-image/pull/3)) can be enabled through configuration on the Temporal server charm.

**Updated:**

Following reviews, I have re-factored this PR to not only include authentication and authorization docs, but to also include wrapper actions for interacting with the OpenFGA store to add the necessary rules, to avoid the use of external clients. This comes in the form of introducing four actions that are able to perform 10 different operations depending on the combination of parameters provided as follows:
1. `add-auth-rule`:
   - `user` and `group`: adds a `user` as a member of a `group`.
   - `group`, `namespace` and `role`: assign the `group` `<role>` access to the `namespace`.
2. `remove-auth-rule`:
   - `user` and `group`: removes a `user` as a member from a `group`.
   -  `group`, `namespace` and `role`: assign the `group` `<role>` access to the `namespace`.
3. `check-auth-rule`:
   - `user` and `group`: checks if a `user` is a member of a `group`.
   - `user`, `namespace` and `role`: checks if a `user` has `<role>` access to a `namespace`.
   - `group`, `namespace` and `role`: checks if members of a `group` have `<role>` access to a `namespace`.
4. `list-auth-rule`:
   - `user`: lists all groups that a `user` is a member of and all namespaces that a `user` has access to.
   - `group`: lists all namespaces that a `group` has access to.
   - `namespace`: lists all groups that have access to this `namespace`.

### CODE REVIEW

File: documentation/how-to/authorization.md
Comment (kian99) [inline documentation/how-to/authorization.md:25]: I think this can go to the stable channel, I've been waiting for another team to use it and give feedback before we do so. Let us know if you're having success with it.

File: documentation/how-to/authorization.md
Comment (kian99) [inline documentation/how-to/authorization.md:206]: It would be nice to have a generic ofga CLI client Snap. I can propose that within our team if it would be useful for you?

**Review by kian99 — APPROVED** [review]
lgtm!

File: documentation/how-to/authorization.md
Comment (kian99) [reply documentation/how-to/authorization.md:206]: It's definitely not ideal giving someone a block of code. But in fact, I think I recall the OpenFGA mentioning that they were working on a CLI.

Comment (kian99) [reply documentation/how-to/authorization.md:206]: Found it! https://github.com/openfga/cli

**Review by TeodorPt — APPROVED** [review]
Nice documentation! Some very small comments and nits.

File: documentation/how-to/authorization.md
Comment (maabujayyab) [inline documentation/how-to/authorization.md:257]: This is not Temporal charm specific and is piercing the abstraction of using OpenFGA here. We either provide a Temoral auth cli interface on top of the generic tuple addition OpenFGA provides that doesn't use OpenFGA's language (i.e. `adduser` or `addgroup`) or completely drop this from the documentation. I can see this relevant in the OpenFGA charm documentation.

**Review by maabujayyab — CHANGES_REQUESTED** [review]
It's always a pleasure to read documentation you write, Ali. I want us to focus what the Temporal charm does and avoid the config exposure.

File: documentation/how-to/authorization.md
Comment (kelkawi-a) [reply documentation/how-to/authorization.md:257]: I agree with your perspective. The way I see it, there are two ways we can approach this to avoid having the user run a program or use a cli tool to add tuples:

1. Add actions to the Temporal charm to add, remove and list tuples. These actions would be specific to the authorization model of Temporal (i.e. you specify user, group and namespace for tuples).
2. Add actions to the OpenFGA charm to add, remove and list tuples. These actions would be generic and would work across different authorization models for multiple applications. You would, however, need a way of specifying the store that you are querying.

These two approaches are not mutually exclusive, so I might be leaning towards the first option to have a Temporal-specific wrapper around these actions for now, while also looking into adding actions to the OpenFGA charm to make this use-case more general for the future.

What do you think?

Comment (gtato) [reply documentation/how-to/authorization.md:257]: To me the the second options seems more appropriate. I don't think temporal should care about adding users and groups (not even as a proxy). We can imagine that the authorization was done centrally and we had no write access to it (which i think it is a common case). You could point to the documentation of the "authorization backend" to add users/permissions.

File: documentation/how-to/authorization.md
Comment (maabujayyab) [inline documentation/how-to/authorization.md:195]: `define writer: [group#member] or admin` - it doesn't make sense for an admin to have read-only access.

**Review by maabujayyab — CHANGES_REQUESTED** [review]

**Review by mertalpt — APPROVED** [review]
LGTM.

File: documentation/how-to/authorization.md
Comment (kelkawi-a) [reply documentation/how-to/authorization.md:25]: I agree, I've tested that the charm works well, and I believe the Contracts team has also tested it.

Comment (kelkawi-a) [reply documentation/how-to/authorization.md:25]: I agree, I've tested that the charm works well.

File: documentation/how-to/authorization.md
Comment (kelkawi-a) [reply documentation/how-to/authorization.md]: Thanks! I've re-factored this PR to also include charm actions that allow the user to interact with the OpenFGA store by adding, removing and checking authorization rules.

File: documentation/how-to/authentication.md
Comment (kelkawi-a) [reply documentation/how-to/authentication.md]: Changed it to the new public repo.

File: documentation/how-to/authentication.md
Comment (kelkawi-a) [reply documentation/how-to/authentication.md:53]: Workers are a type of Temporal client, but I made the distinction here in that Temporal clients are what is used to trigger workflows, while workers are what is used to execute these workflows. I've made the change to `worker operator instances`.

File: documentation/how-to/authorization.md
Comment (kelkawi-a) [reply documentation/how-to/authorization.md]: I've refactored this PR and implemented the necessary actions to achieve this. However, it is slightly different to what you described, as OpenFGA deals in tuples, so you cannot add a user or add a group, but rather add the relation between a user and a group. All mentioned operations can be achieved using the four actions that have been implemented.

File: documentation/how-to/authorization.md
Comment (mertalpt) [inline documentation/how-to/authorization.md:30]: Where is it written how to deploy all the other charms in the model? It might be nice to put a reference to how we got to his point.

File: documentation/index.md
Comment (mertalpt) [inline documentation/index.md:44]: Don't you have the files for these now, why are they still `TODO`?

File: src/relations/openfga.py
Comment (mertalpt) [inline src/relations/openfga.py:446]: If you are using Python `3.10<=`  check out `match` statement, this is the perfect place for it.

**Review by mertalpt — APPROVED** [review]
LGTM.

File: documentation/index.md
Comment (kelkawi-a) [reply documentation/index.md:44]: Once the PR is reviewed and approved, i'll be creating Discourse pages and replacing these `TODO`s with links to these pages.

File: documentation/how-to/authorization.md
Comment (kelkawi-a) [reply documentation/how-to/authorization.md:30]: The first section of this page mentions this:

> Enabling authorization requires that you have an active [Charmed Temporal K8s Operator](https://discourse.charmhub.io/t/charmed-temporal-k8s-tutorial-introduction/11777) deployed as described in the tutorial.

**Review by AmberCharitos — APPROVED** [review]
LGTM, nice work!

**Review by maabujayyab — APPROVED** [review]
LGTM! This is shaping up quite well.

File: documentation/how-to/authorization.md
Comment (kelkawi-a) [reply documentation/how-to/authorization.md:193]: Good suggestion. For now, there is no mechanism for adding users in bulk, but it's something we can consider incorporating into the actions in the future.

File: tests/integration/test_auth.py
Comment (kelkawi-a) [reply tests/integration/test_auth.py:114]: There's currently no tests being done with an authenticated user/service account as it requires further work and including some credentials as Github secrets. I'll create a story for that specifically.

**Pull request merged**
